app/helpers/async_request/application_helper.rb

```diff
@@ -1,15 +1,44 @@
 module AsyncRequest
   module ApplicationHelper
-    def execute_async(worker_class, *params)
+    def execute_async(worker_class, *worker_params)
       raise ArgumentError if worker_class.nil?
+      @worker_params = worker_params
+      @worker_class = worker_class
       job = Job.create!(
         worker: worker_class,
-        params: params,
+        params: set_worker_params,
         status: Job.statuses[:waiting],
         uid: SecureRandom.uuid
       )
       JobProcessor.perform_async(job.id, job.worker)
       job.uid
     end
+
+    def set_extra_data(new_data)
+      @extra_data = new_data
+    end
+
+    def add_extra_data(new_data)
+      extra_data.merge!(new_data)
+    end
+
+    private
+
+    def extra_data
+      @extra_data ||= {}
+    end
+
+    def set_worker_params 
+      return @worker_params unless send_extra_data?
+      params_and_extra_data
+    end
+
+    def send_extra_data?
+      !extra_data.empty? && @worker_class.respond_to?(:handle_extra_data?) && @worker_class.handle_extra_data?
```

Comment:
Maybe `extra_data.present?` instead of not empty 🤔 
The `respond_to` check is to avoid calling a nonexistent method, right?
Where is `handle_extra_data?` being defined?

Reply:
Yes, that's better way better, thanks!
The `handle_extra_data?` is a class method defined by each worker (or the base worker). This means that after checking that `extra_data` is not empty, we ask the worker's class if it is capable of answering if it can handle extra data. If it has the method defined, then we ask the worker's class if it wants to handle the extra data. 
If you want we can talk about this through a call!